File: Python DSA/Trees/path_sum.py
"""
PATH SUM - Easy
Check if tree has root-to-leaf path with target sum.

Logical Thinking:
1. Subtract current value from target
2. Check if leaf node with sum 0
3. Recurse on children

Logical Thinking:
def has_path_sum(root, target_sum):
    if not root:
        return False
    if not root.left and not root.right:
        return root.val == target_sum
    target_sum -= root.val
    return has_path_sum(root.left, target_sum) or has_path_sum(root.right, target_sum)
"""

# TODO: Implement solution


# Test cases
if __name__ == "__main__":
    print(f"Testing Path Sum...")
    print("Add your test cases here")

File: DSA/Binary Tree/Java/PathSum.java
// LeetCode 112: Path Sum
// Given the root of a binary tree and an integer targetSum, return true if the tree has a root-to-leaf path such that adding up all the values along the path equals targetSum.

public class PathSum {
    public boolean hasPathSum(TreeNode root, int targetSum) {
        if (root == null) {
            return false;
        }
        
        // If it's a leaf node, check if the remaining sum equals the node's value
        if (root.left == null && root.right == null) {
            return targetSum == root.val;
        }
        
        // Recursively check left and right subtrees with updated target sum
        int remainingSum = targetSum - root.val;
        return hasPathSum(root.left, remainingSum) || hasPathSum(root.right, remainingSum);
    }
    
    public static void main(String[] args) {
        PathSum solution = new PathSum();
        
        // Create test tree: [5,4,8,11,null,13,4,7,2,null,null,null,1]
        TreeNode root = new TreeNode(5);
        root.left = new TreeNode(4);
        root.right = new TreeNode(8);
        root.left.left = new TreeNode(11);
        root.right.left = new TreeNode(13);
        root.right.right = new TreeNode(4);
        root.left.left.left = new TreeNode(7);
        root.left.left.right = new TreeNode(2);
        root.right.right.right = new TreeNode(1);
        
        int targetSum = 22;
        System.out.println("Has path sum " + targetSum + ": " + solution.hasPathSum(root, targetSum));
    }
}